pkg/dockerregistry/server/repositorymiddleware.go

```diff
@@ -416,68 +339,84 @@ func (r *repository) Put(manifest *schema1.SignedManifest) error {
-			return statusErr
+			return "", statusErr
 		}
 
 		// try to create the ISM again
 		if err := r.registryOSClient.ImageStreamMappings(r.namespace).Create(&ism); err != nil {
 			if quotautil.IsErrorQuotaExceeded(err) {
 				context.GetLogger(r.ctx).Errorf("denied a creation of ImageStreamMapping: %v", err)
-				return distribution.ErrAccessDenied
+				return "", distribution.ErrAccessDenied
 			}
 			context.GetLogger(r.ctx).Errorf("error creating ImageStreamMapping: %s", err)
-			return err
+			return "", err
 		}
 	}
 
-	// Grab each json signature and store them.
-	signatures, err := manifest.Signatures()
-	if err != nil {
-		return err
-	}
+	return dgst, nil
+}
 
-	for _, signature := range signatures {
-		if err := r.Signatures().Put(dgst, signature); err != nil {
-			context.GetLogger(r.ctx).Errorf("error storing signature: %s", err)
+// fillImageWithMetadata fills a given image with metadata. It also corrects layer sizes with blob sizes. Newer
+// Docker client versions don't set layer sizes in the manifest at all. Origin master needs correct layer
+// sizes for proper image quota support. That's why we need to fill the metadata in the registry.
+func (r *repository) fillImageWithMetadata(manifest distribution.Manifest, image *imageapi.Image) error {
+	if deserializedManifest, ok := manifest.(*schema2.DeserializedManifest); ok {
+		configBytes, err := r.Blobs(r.ctx).Get(r.ctx, deserializedManifest.Config.Digest)
+		if err != nil {
+			context.GetLogger(r.ctx).Errorf("failed to get image config %s: %v", deserializedManifest.Config.Digest.String(), err)
 			return err
 		}
+		image.DockerImageConfig = string(configBytes)
 	}
 
-	return nil
-}
+	if signedManifest, ok := manifest.(*schema1.SignedManifest); ok {
+		signatures, err := signedManifest.Signatures()
+		if err != nil {
+			return err
+		}
+
+		for _, signDigest := range signatures {
+			image.DockerImageSignatures = append(image.DockerImageSignatures, signDigest)
+		}
+	}
 
-// fillImageWithMetadata fills a given image with metadata. Also correct layer sizes with blob sizes. Newer
-// Docker client versions don't set layer sizes in the manifest at all. Origin master needs correct layer
-// sizes for proper image quota support. That's why we need to fill the metadata in the registry.
-func (r *repository) fillImageWithMetadata(manifest *schema1.SignedManifest, image *imageapi.Image) error {
 	if err := imageapi.ImageWithMetadata(image); err != nil {
 		return err
 	}
 
+	refs := manifest.References()
+
 	layerSet := sets.NewString()
 	size := int64(0)
 
 	blobs := r.Blobs(r.ctx)
 	for i := range image.DockerImageLayers {
 		layer := &image.DockerImageLayers[i]
-		// DockerImageLayers represents manifest.Manifest.FSLayers in reversed order
-		desc, err := blobs.Stat(r.ctx, manifest.Manifest.FSLayers[len(image.DockerImageLayers)-i-1].BlobSum)
+		// DockerImageLayers represents manifest.Manifest.Layers in reversed order
```

Comment:
Doesn't the schema2 manifest already have this info (mediaType and size)?  It doesn't seem like we need this if we already have that info.